src/middlewared/middlewared/plugins/vm.py

```diff
@@ -1,487 +1,651 @@
+        if self.vm_data['bootloader'] == 'GRUB' and not any(
+            isinstance(d, RAW) and d.data['attributes'].get('boot') for d in self.devices
+        ):
+            raise CallError(f'Unable to find boot devices for {self.libvirt_domain_name} domain')
+
+        if len([d for d in self.devices if isinstance(d, VNC)]) > 1:
+            raise CallError('Only one VNC device per VM is supported')
+
+        for device in self.devices:
+            device.pre_start_vm()
+
+        if self.domain.create() < 0:
+            raise CallError(f'Failed to boot {self.vm_data["name"]} domain')
+
+        try:
+            for device in self.devices:
+                device.post_start_vm()
+        finally:
+            # We initialize this when we are certain that the VM has indeed booted
+            self.stop_devices_thread = threading.Thread(
+                name=f'post_stop_devices_{self.libvirt_domain_name}', target=self.run_post_stop_actions
+            )
+
+    def _before_stopping_checks(self):
+        if not self.domain.isActive():
+            raise CallError(f'{self.libvirt_domain_name} domain is not active')
+
+    def initiate_post_stop_thread(self):
+        with self.stop_lock:
+            # Following ensures that the thread has not already completed it's execution
+            if self.stop_devices_thread and not self.stop_devices_thread.ident:
+                self.stop_devices_thread.start()
+
+    def run_post_stop_actions(self):
+        while self.status()['state'] == 'RUNNING':
+            time.sleep(5)
+
+        for device in self.devices:
+            device.post_stop_vm()
+
+    def stop(self, shutdown_timeout=90):
+        self._before_stopping_checks()
+
+        for device in self.devices:
+            device.pre_stop_vm()
+
+        self.domain.shutdown()
+
+        # We wait for timeout seconds before initiating post stop activities for the vm
+        # This is done because the shutdown call above is non-blocking
+        while shutdown_timeout > 0 and self.status()['state'] == 'RUNNING':
+            shutdown_timeout -= 5
+            time.sleep(5)
+
+        self.initiate_post_stop_thread()
+
+    def restart(self, vm_data=None, shutdown_timeout=90):
+        self.stop(shutdown_timeout)
+
+        # We don't wait anymore because during stop we have already waited for the VM to shutdown cleanly
+        if self.status()['state'] == 'RUNNING':
+            # In case domain stopped between this time
+            with contextlib.suppress(libvirt.libvirtError):
+                self.poweroff()
+
+        self.start(vm_data)
+
+    def poweroff(self):
+        self._before_stopping_checks()
+
+        for device in self.devices:
+            with contextlib.suppress(Exception):
+                device.pre_stop_vm()
+
+        self.domain.destroy()
+
+        self.initiate_post_stop_thread()
+
+        with contextlib.suppress(Exception):
+            self.stop_devices_thread.join()
```

Comment:
It has two names: `post_stop_thread` and `stop_devices_thread`, we need to choose one